### ISSUE TITLE
Add Gradient scaler

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#832]
With the addition of `AMP<F>` dtype, we also need to add gradient scaling, which is commonly used with AMP training.

I think the frontend interface could look something like:
```rust
let mut scaler = GradientScaler { ... }; // similar fields to pytorch scalar

// this would do both parts that you have to do in pytorch now:
// 1. would scale the loss by the correct value
// 2. would unscale the gradients before returning them
grads = scaler.scaled_backward(loss);
```

We may have to add some methods to Gradients to support scaling them.

_Originally posted by @coreylowman in https://github.com/coreylowman/dfdx/issues/424#issuecomment-1650162800_

**Comment by coreylowman** [comment]
Posting pytorch documentation here: https://pytorch.org/docs/stable/amp.html#gradient-scaling
